### PR TITLE
Generate type mappings in the compiled model.

### PR DESCRIPTION
**Pull request by AndriySvyryd**
All type mapping types now need to be public and have a public static property `Default` of its own type.
Annotate NTS as not supported with NativeAOT.

Part of #29761

### CODE REVIEW

**Review by bricelam — COMMENTED** [review]
LGTM

File: src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs [hunk 1]
```diff
@@ -867,31 +836,6 @@ private void Create(IEntityType entityType, CSharpRuntimeAnnotationCodeGenerator
                 .Append("()");
         }
 
-        var jsonValueReaderWriterType = (Type?)property[CoreAnnotationNames.JsonValueReaderWriterType];
```
Comment (ajcvickers) [inline src/EFCore.Design/Scaffolding/Internal/CSharpRuntimeModelCodeGenerator.cs:870]: 👍

File: src/EFCore.Relational/Storage/ByteArrayTypeMapping.cs [hunk 2]
```diff
@@ -23,6 +23,14 @@ namespace Microsoft.EntityFrameworkCore.Storage;
 /// </remarks>
 public class ByteArrayTypeMapping : RelationalTypeMapping
 {
+    /// <summary>
+    ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
+    ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
+    ///     any release. You should only use it directly in your code with extreme caution and knowing that
+    ///     doing so can result in application failures when updating to a new Entity Framework Core release.
+    /// </summary>
+    public static ByteArrayTypeMapping Default { get; } = new("binary");
```
Comment (ajcvickers) [inline src/EFCore.Relational/Storage/ByteArrayTypeMapping.cs:32]: These seem like provider-specific defaults but in the Relational assembly.

Comment (AndriySvyryd) [reply src/EFCore.Relational/Storage/ByteArrayTypeMapping.cs:32]: We need some default. Do you have a different proposal?

File: src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs [hunk 3]
```diff
@@ -895,6 +896,49 @@ ITuple tuple
                          && tupleType.Namespace == "System"
                     => HandleValueTuple(tuple),
 
+                IEqualityComparer c
```
Comment (roji) [inline src/EFCore.Design/Query/Internal/LinqToCSharpSyntaxTranslator.cs:899]: In the future we may want to expose a hook to allow registering more "well-known" constants.

File: src/EFCore/Design/Internal/CSharpRuntimeAnnotationCodeGenerator.cs [hunk 4]
```diff
@@ -336,10 +339,261 @@ protected virtual void AddNamespace(Type type, ISet<string> namespaces)
...
+        AddNamespace(converter.ProviderClrType, parameters.Namespaces);
+
+        mainBuilder
+            .Append("new ValueConverter<")
+            .Append(codeHelper.Reference(converter.ModelClrType))
+            .Append(", ")
+            .Append(codeHelper.Reference(converter.ProviderClrType))
+            .AppendLine(">(")
+            .IncrementIndent()
+            .Append(codeHelper.Expression(converter.ConvertToProviderExpression, parameters.Namespaces))
```
Comment (roji) [inline src/EFCore/Design/Internal/CSharpRuntimeAnnotationCodeGenerator.cs:372]: Super cool!

**Review by roji — COMMENTED** [review]
LGTM for all the LINQ->C# changes, it's really nice to see this!

We'll see when people start using it exactly how many C# features aren't actually supported by it :grimacing:

**Pull request merged**